src/woo_publications/publications/api/viewsets.py

```diff
@@ -65,6 +77,54 @@ def perform_create(self, serializer):
             build_absolute_uri=self.request.build_absolute_uri,
         )
 
+    @extend_schema(
+        summary=_("Upload file part"),
+        description=_(
+            "Send the binary data for a file part to perform the actual file upload. "
+            "The response data of the document create endpoints returns the list of "
+            "expected file parts, pointing to this endpoint. The client must split "
+            "the binary file in the expected part sizes and then upload each chunk "
+            "individually.\n\n"
+            "Once all file parts for the document are received, the document will be "
+            "automatically unlocked in the Documents API and ready for use.\n\n"
+            "**NOTE** this endpoint expects `multipart/form-data` rather than JSON to "
+            "avoid the base64 encoding overhead."
+        ),
+        responses={200: DocumentStatusSerializer},
+    )
+    @action(
+        detail=True,
+        methods=["put"],
+        serializer_class=FilePartSerializer,
+        parser_classes=(MultiPartParser,),
+        url_path="bestandsdelen/<uuid:part_uuid>",
+        url_name="filepart-detail",
+    )
+    def file_part(self, request: Request, part_uuid: UUID, *args, **kwargs) -> Response:
+        document: Document = self.get_object()
+        serializer = FilePartSerializer(
+            data=request.data,
+            context={"request": request, "view": self},
+        )
+        serializer.is_valid(raise_exception=True)
+
+        file = serializer.validated_data["inhoud"]
+        try:
+            is_completed = document.upload_part_data(uuid=part_uuid, file=file)
+        except RequestException as exc:
+            # we can only handle HTTP 400 responses
+            if (_response := exc.response) is None:
+                raise
+            if _response.status_code != 400:
+                raise
+            # XXX: should we transform these error responses?
```

Comment:
A generic "oops" error might be nicer yeah, although I am personally fine with this myself.